File: sdks/go/cmd/elydora/plugins/gemini_io.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
)

func geminiConfigurationDirectory() (string, error) {
	configured := os.Getenv("GEMINI_CLI_HOME")
	if configured == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("resolve home directory: %w", err)
		}
		configured = home
	}
	return filepath.Join(configured, ".gemini"), nil
}

func geminiSettingsPath() (string, error) {
	directory, err := geminiConfigurationDirectory()
	if err != nil {
		return "", err
	}
	return filepath.Join(directory, geminiConfigFile), nil
}

func readGeminiDocument() (*geminiDocument, error) {
	filePath, err := geminiSettingsPath()
	if err != nil {
		return nil, err
	}
	if _, err := managedPhysicalDirectoryExists(
		filepath.Dir(filePath),
		"Gemini CLI configuration directory",
	); err != nil {
		return nil, err
	}
	snapshot, err := readManagedFile(
		filePath,
		"Gemini CLI user settings",
		maxManagedSourceBytes,
	)
	if err != nil {
		return nil, err
	}
	if snapshot == nil {
		return createGeminiDocument(filePath)
	}
	return parseGeminiDocument(true, filePath, snapshot.contents)
}

func prepareRenderedGeminiChange(
	rendered *geminiRenderedDocument,
) (*fileChange, error) {
	if rendered == nil || !rendered.changed {
		return nil, nil
	}
	return prepareSourceChange(
		rendered.document.filePath,
		"Gemini CLI user settings",
		rendered.document.raw,
		rendered.document.exists,
		rendered.next,
		0600,
		rendered.remove,
	)
}

func writeGeminiChanges(
	changes []*fileChange,
	label string,
	rename renameFunc,
	runtimeRoot string,
	agentDirectory string,
	settingsDirectory string,
) error {
	hasChanges := false
	for _, change := range changes {
		if change != nil {
			hasChanges = true
			break
		}
	}
	if !hasChanges {
		return nil
	}
	if agentDirectory != "" {
		if err := EnsurePrivateDirectory(runtimeRoot); err != nil {
			return err
		}
		if err := EnsurePrivateDirectory(agentDirectory); err != nil {
			return err
		}
	}
	if err := ensureManagedDirectory(
		settingsDirectory,
		"Gemini CLI configuration directory",
	); err != nil {
		return err
	}
	return writeChanges(changes, label, rename)
}

File: sdks/node/src/plugins/gemini-io.ts
import os from 'node:os';
import path from 'node:path';
import {
  AGENT_KEY,
  AUDIT_OPTIONS,
  CONFIG_FILE,
  GUARD_OPTIONS,
  type GeminiRuntimeContract,
} from './gemini-contract.js';
import {
  createGeminiDocument,
  parseGeminiDocument,
  type GeminiDocument,
  type RenderedGeminiDocument,
} from './gemini-config.js';
import { inspectPhysicalDirectory, readPhysicalFile } from './managed-files.js';
import { managedRuntimeFilesExist } from './managed-runtime-status.js';
import {
  commitManagedTransaction,
  prepareManagedFileChange,
} from './managed-transaction.js';

export function geminiConfigurationDirectory(): string {
  const configuredHome = process.env.GEMINI_CLI_HOME;
  const home = configuredHome ? configuredHome : os.homedir();
  return path.join(home, '.gemini');
}

export function geminiSettingsPath(): string {
  return path.join(geminiConfigurationDirectory(), CONFIG_FILE);
}

export async function readGeminiDocument(): Promise<GeminiDocument> {
  const filePath = geminiSettingsPath();
  await inspectPhysicalDirectory(path.dirname(filePath), 'Gemini CLI configuration directory');
  const snapshot = await readPhysicalFile(filePath, 'Gemini CLI user settings');
  return snapshot
    ? parseGeminiDocument({
      exists: true,
      filePath,
      raw: snapshot.contents,
    })
    : createGeminiDocument(filePath);
}

export async function writeGeminiDocument(rendered: RenderedGeminiDocument): Promise<void> {
  if (!rendered.changed) return;
  const change = await prepareManagedFileChange({
    filePath: rendered.document.filePath,
    label: 'Gemini CLI user settings',
    next: rendered.next,
    mode: 0o600,
    expectedSource: rendered.document.exists ? rendered.document.raw : undefined,
    verifyExpectedSource: true,
  });
  if (!change) return;
  await commitManagedTransaction({
    displayName: 'Gemini CLI',
    operation: 'uninstall',
    directories: [{
      path: path.dirname(rendered.document.filePath),
      label: 'Gemini CLI configuration directory',
    }],
    changes: [change],
  });
}

export async function geminiRuntimeFilesExist(
  contracts: readonly GeminiRuntimeContract[],
): Promise<boolean> {
  for (const contract of contracts) {
    const exists = await managedRuntimeFilesExist(contract, AGENT_KEY, {
      guardOptions: GUARD_OPTIONS,
      auditOptions: AUDIT_OPTIONS,
    });
    if (exists) return true;
  }
  return false;
}
